File: src/Payment.hs
{-# LANGUAGE TupleSections #-}

module Payment
  ( BillingAmount(..)
  , SharingAmount(..)
  , FractionAmount(..)
  , AdjustingUnitCount(..)
  , AdjustingUnitAmount(..)
  , PaymentAmount(..)
  , PaymentSection(..)
  , PaymentError(..)
  , DivideNumber(..)
  , calculate
  , divide
  , adjustByUnit
  , sharing
  , (<++)
  ) where

import Fundamental
import Data.List (nub)

--
-- Types
--

newtype BillingAmount = BillingAmount Money deriving (Show, Eq)

newtype SharingAmount = SharingAmount Money deriving (Show, Eq)

newtype FractionAmount = FractionAmount Money deriving (Show, Eq)

newtype AdjustingUnitCount = AdjustingUnitCount Int deriving (Show, Eq)

newtype AdjustingUnitAmount = AdjustingUnitAmount Money deriving (Show, Eq)

newtype PaymentAmount = PaymentAmount Money deriving (Show, Eq)

newtype DivideNumber = DivideNumber Int deriving (Show, Eq)

--
-- Sections
--

data PaymentSection = More | Normal | Less deriving (Show, Eq)

--
-- Errors
--

data PaymentError = UnsharableAdjustingUnitAmount | TooLargeAdjustingUnitAmount deriving (Show, Eq)

--
-- Functions
--

calculate :: BillingAmount -> [PaymentSection] -> AdjustingUnitAmount -> Either PaymentError ([(PaymentSection, PaymentAmount)], FractionAmount)
calculate billingAmount paymentSections adjustingUnitAmount = do
  let divideNumber = DivideNumber $ length paymentSections
  (sharingAmount, fractionAmount) <- divide billingAmount divideNumber adjustingUnitAmount
  let units = adjustByUnit paymentSections
  (, fractionAmount) <$> sharing sharingAmount adjustingUnitAmount divideNumber units
  

divide :: BillingAmount -> DivideNumber -> AdjustingUnitAmount -> Either PaymentError (SharingAmount, FractionAmount)
divide (BillingAmount (Money ba)) (DivideNumber dn) (AdjustingUnitAmount (Money aua)) = case ba `div` dn `div` aua * dn * aua of
  0       -> Left  UnsharableAdjustingUnitAmount
  sharing -> Right (SharingAmount $ Money sharing, FractionAmount $ Money $ ba - sharing)


adjustByUnit :: [PaymentSection] -> [(PaymentSection, AdjustingUnitCount)]
adjustByUnit paymentSections = let
  (pointer, adjuster) = if More `elem` paymentSections then morePresent else moreMissing

  plus = sum . filter (> 0) . map pointer $ paymentSections
  minus = sum . filter (< 0) . map pointer $ paymentSections
  lcm' = lcm plus minus

  in map (adjuster lcm' plus minus) $ nub paymentSections
    where
      morePresent = (point, adjust)
        where
          point More   = 1
          point Normal = -1
          point Less   = -2

          adjust lcm' plus _     More   = (More,   AdjustingUnitCount $ lcm' `div` plus)
          adjust lcm' _    minus Normal = (Normal, AdjustingUnitCount $ lcm' `div` minus)
          adjust lcm' _    minus Less   = (Less,   AdjustingUnitCount $ lcm' `div` minus * 2)

      moreMissing = (point, adjust)
        where
          point Normal = 1
          point Less   = -1

          adjust lcm' plus _     Normal = (Normal, AdjustingUnitCount $ lcm' `div` plus)
          adjust lcm' _    minus Less   = (Less,   AdjustingUnitCount $ lcm' `div` minus)


sharing :: SharingAmount -> AdjustingUnitAmount -> DivideNumber -> [(PaymentSection, AdjustingUnitCount)] -> Either PaymentError [(PaymentSection, PaymentAmount)]
sharing (SharingAmount (Money sa)) adjustingUnitAmount (DivideNumber dn) units = let
  ave = Money $ sa `div` dn
  fixed = map (mapR (fix ave adjustingUnitAmount)) units
  in if all (>=0) $ map (snd . mapR asInt) fixed then Right fixed  else Left TooLargeAdjustingUnitAmount
    where
      fix :: Money -> AdjustingUnitAmount -> AdjustingUnitCount -> PaymentAmount
      fix ave (AdjustingUnitAmount aua) (AdjustingUnitCount auc) = PaymentAmount $ aua $*$ auc $+$ ave

      asInt :: PaymentAmount -> Int
      asInt (PaymentAmount (Money i)) = i
      

(<++) :: PaymentAmount -> FractionAmount -> PaymentAmount
PaymentAmount m1 <++ FractionAmount m2 = PaymentAmount $ m1 $+$ m2

File: 2020-01-python/payment.py
def calculate(billingAmount, paymentSections, adjustingUnitAmount):
    divideNumber = len(paymentSections)
    (sharingAmount, fractionAmount) = divide(billingAmount, divideNumber, adjustingUnitAmount)
    print 'divide: ',
    print (sharingAmount, fractionAmount)
    units = adjustByUnit(paymentSections)
    print 'units: ',
    print units
    return (sharing(sharingAmount, adjustingUnitAmount, divideNumber, units), fractionAmount)


def divide(billingAmount, divideNumber, adjustingUnitAmount):
    sharing = billingAmount / divideNumber / adjustingUnitAmount * divideNumber * adjustingUnitAmount
    if sharing == 0:
        raise BaseException('unsharable adjusting unit amount')
    else:
        return (sharing, billingAmount - sharing)


def adjustByUnit(paymentSections):
    def get_lcm(x, y):
        import fractions
        return (x * y) // fractions.gcd(x, y)

    def morePresent():
        def point(paymentSection):
            return {'M': 1, 'N': -1, 'L': -2}[paymentSection]
        
        def adjust(lcm, plus, minus, paymentSection):
            return {'M': ('M', lcm / plus), 'N': ('N', lcm / minus), 'L': ('L', lcm / minus * 2)}[paymentSection]

        return (point, adjust)

    def moreMissing():
        def point(paymentSection):
            return {'N': 1, 'L': -1}[paymentSection]
        
        def adjust(lcm, plus, minus, paymentSection):
            return {'N': ('N', lcm / plus), 'L': ('L', lcm / minus)}[paymentSection]

        return (point, adjust)

    (pointer, adjuster) = morePresent() if 'M' in paymentSections else moreMissing()

    plus  = sum(filter(lambda n: n > 0, map(pointer, paymentSections)))
    minus = sum(filter(lambda n: n < 0, map(pointer, paymentSections)))
    lcm = get_lcm(plus, minus)

    return [adjuster(lcm, plus, minus, paymentSection) for paymentSection in list(set(paymentSections))]


def sharing(sharingAmount, adjustingUnitAmount, divideNumber, units):
    def fix(ave, adjustingUnitAmount, adjustingUnitCount):
        return adjustingUnitAmount * adjustingUnitCount + ave

    ave = sharingAmount / divideNumber
    fixed = [(paymentSection, fix(ave, adjustingUnitAmount, adjustingUnitCount)) for (paymentSection, adjustingUnitCount) in units]

    if all([paymentAmount >= 0 for (paymentSection, paymentAmount) in fixed]):
        return fixed
    else:
        raise BaseException('too large adjusting unit amount')


# print divide(1000, 3, 100) # 900, 100
# print divide(1199, 3, 100) # 900, 299
# print divide(1200, 3, 500) # unsharable adjusting unit amount

# print adjustByUnit(['M', 'M', 'M', 'N', 'L', 'L']) #  5, -3, -6
# print adjustByUnit(['M', 'N'])                     #  1, -1
# print adjustByUnit(['N', 'L'])                     #  1, -1

# print sharing(  900,  100, 3, [('M', 3), ('N', -1), ('L', -2)]) #   600,  200,  100
# print sharing(60000, 1000, 6, [('M', 5), ('N', -3), ('L', -6)]) # 15000, 7000, 4000
# print sharing( 6000, 1000, 6, [('M', 5), ('N', -3), ('L', -6)]) # too large adjusting unit amount

# print calculate(  900, ['M', 'N', 'L'], 100) # 600, 200, 100,  0
# print calculate(  990, ['M', 'N', 'L'], 100) # 600, 200, 100, 90
# print calculate(65000, ['M', 'M', 'M', 'N', 'L', 'L'], 500) # 13000, 9000, 7500
